File: compiler/RangerAppFunctionDesc.clj
(
    ( CreateClass RangerAppFunctionDesc
        (
            (def name:string "")
            
            (def node:CodeNode @weak(true))            
            (def nameNode:CodeNode @weak(true))

            (def params:[RangerAppParamDesc])
            (def return_value:RangerAppParamDesc)

            (def is_method:boolean false)
            (def is_static false)
            (def container_class:RangerAppClassDesc)

            (def refType:RangerNodeRefType RangerNodeRefType.NoType)            

            ; function body AST and string
            (def body_ast:CodeNode)
            (def body_str:string)

        )
    )
)

File: versions/java/RangerAppFunctionDesc.java
import java.util.Optional;
import java.util.*;

class RangerAppFunctionDesc extends RangerAppParamDesc { 
  public Optional<CodeNode> fnBody = Optional.empty();
  public ArrayList<RangerAppParamDesc> params = new ArrayList<RangerAppParamDesc>();
  public Optional<RangerAppParamDesc> return_value = Optional.empty()     /** note: unused */;
  public boolean is_method = false     /** note: unused */;
  public boolean is_static = false;
  public Optional<RangerAppClassDesc> container_class = Optional.empty()     /** note: unused */;
  public Optional<RangerAppWriterContext> fnCtx = Optional.empty();
  
  public boolean isClass() {
    return false;
  }
  
  public boolean isProperty() {
    return false;
  }
}
